File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null && list2 == null) return list1;
        if (list1 == null) return list2;
        if (list2 == null) return list1;

        ListNode head = null;
        ListNode node1 = list1;
        ListNode node2 = list2;
        if (node1.val <= node2.val) {
            head = node1;
            node1 = node1.next;
        } else {
            head = node2;
            node2 = node2.next;
        }
        ListNode res = head;

        while (node1 != null && node2 !=null) {
            ListNode temp = head;
            if (node1.val <= node2.val) {
                head.next = node1;
                node1 = node1.next;
            } else {
                head.next = node2;
                node2 = node2.next;
            }
            head = head.next;
        }

        if (node1 != null) {
            head.next = node1;
        } else {
            head.next = node2;
        }

        return res;
    }
}

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head: ListNode | null): ListNode {
        const dum = new ListNode()
        dum.next = head
        let cur = head;

        while (cur && cur.next) {
            let sec = cur.next
            cur.next = sec.next
            sec.next = dum.next
            dum.next = sec
        }

        return dum.next
    }
}

/**
 * Space: O(1)
 * Time: O(n)
 * 
 * Dry run:
 * head = [0, 1, 2, 3], dum.next = 0, cur = 0
 * sec  | cur.next  | cur   | sec.next  | dum.next  | head
 * 1    | 2         | 2     | 0         | 1         | 1, 0, 2, 3
 * 3    | 
 * 
 */
